pkg/util/hardware/hw_utils.go

```diff
@@ -59,6 +101,65 @@ func ParseCPUSetLine(cpusetLine string) (cpusList []int, err error) {
 	return
 }
 
+func ParseCPUSetFile(file string) ([]int, error) {
+	line, err := util.ScanLine(file)
+	if err != nil {
+		return nil, err
+	}
+	cpusList, err := ParseCPUSetLine(line)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse cpuset file: %v", err)
+	}
+	return cpusList, nil
+}
+
+func GroupCPUThreads(topo CPUTopology) ([]int, error) {
+	cpus, err := topo.GetCPUs()
+	if err != nil {
+		return nil, err
+	}
+
+	cpuSet := NewCPUSet(cpus)
+	res := make([]int, 0, len(cpus))
+
+	appendAndFinish := func(cpu int) bool {
+		res = append(res, cpu)
+		cpuSet.Remove(cpu)
+		return cpuSet.Empty()
+	}
+
+	for _, cpu := range cpus {
+		if !cpuSet.Has(cpu) {
+			continue
+		}
+
+		if appendAndFinish(cpu) {
+			return res, nil
+		}
+
+		siblings, err := topo.GetCPUSiblings(cpu)
+		if err != nil {
+			log.DefaultLogger().Reason(err).Error("Error while reading CPU topology")
+			return cpus, nil
+		}
+		for _, sibling := range siblings {
+			if !cpuSet.Has(sibling) {
+				continue
+			}
+
+			if appendAndFinish(sibling) {
+				return res, nil
+			}
+		}
+	}
+	return res, nil
```

Comment:
These are still 2 loops... I'd prefer just to prepare a map ahead of time where you will have all siblings. 
Then instead of looping, you would just get the right  sibling (is it exist) from the map in O(1)

Perhaps, take another look at the example I've provided earlier?

Reply:
Hm.... I am not sure I understand how to avoid the loop here. Its `1 to many` relation in general (e.g. I can have a VM with 3 or more threads pre core). While in the example the assumption is `1 to 1`: `siblingCPUs := map[int]int{0:2,2:0,1:13,13:1,4:15,15:4}`. Or is it my misunderstanding?

Reply:
Right, however, this just provides you a list of siblings per cpu id, so for cpu 0 you would get a list of itself and its sibling so: 0, 4
and for cpu id 4 you will get the same.
$ cat  /sys/devices/system/cpu/cpu4/topology/core_cpus_list
0,4
$ cat  /sys/devices/system/cpu/cpu0/topology/core_cpus_list
0,4

you could reduce it to map of cpu id and its sibling so: {0:4, 4:0} 
The benefits from my point of view are:
  1. no second loop during lookup
  2. you don't need to read files when you are creating a vm, but you can have all of this mapping done ahead of time
  3. when no hyperthreading is enabled on the node, this map will just be empty

Does it make sense?

Reply:
Sorry, still I am not sure I follow...

> no second loop during lookup

We can have a usecase when there are 4 threads on a single core. E.g. assuming we have kubevirt running on a node with 2 cores and 4 threads per core (8 logical CPUs). The siblings map will look like
```go
//$ cat /sys/devices/system/cpu/cpu0/topology/core_cpus_list
//0,2,4,6
//$ cat /sys/devices/system/cpu/cpu1/topology/core_cpus_list
//1,3,5,7
siblings := map[int][]int{
	0: {0, 2, 4, 6}, 2: {0, 2, 4, 6}, 4: {0, 2, 4, 6}, 6: {0, 2, 4, 6},
	1: {1, 3, 5, 7}, 3: {1, 3, 5, 7}, 5: {1, 3, 5, 7}, 7: {1, 3, 5, 7},
}
```
Then in order to check all the siblings we need to iterate in a loop over the corresponding list. E.g. for `CPU 0` -> `{0, 2, 4, 6}` (we can skip `0` as its the actual CPU we are checking). That is exactly what is done in the current implementation.

> you don't need to read files when you are creating a vm, but you can have all of this mapping done ahead of time

If I understand correctly the `virt-launcher` pod is run only when creating a vm. Hence I cannot know which CPUs are assigned to that pod ahead of time. That info is also provided via sysfs `/sys/fs/cgroup/cpuset/cpuset.cpus`. Or are you suggesting to create a siblings map for *all* the CPUs available on the node and keep it as _cache_? (hm... that might be an option actually).

> when no hyperthreading is enabled on the node, this map will just be empty

The same as now: with no hyperthreading (or 1 thread per core) the siblings list will be empty (or to be more precise it will contain the CPU itself therefore it will be skipped).

Reply:
> We can have a usecase when there are 4 threads on a single core. E.g. assuming we have kubevirt running on a node with 2 cores and 4 threads per core (8 logical CPUs). The siblings map will look like

4 threads per core? What CPU/architecture are we talking about?

Reply:
Probably on a bare-metal node its indeed max 2 hyperthreads. But it can also be a VM node and its possible to configure the topology for a VM with 4 threads per core. I assume its a valid usecase. Apart from that in general the topology interface in sysfs does not say that we have only up to 2 sibling threads. There might be more and that is why sysfs provides that data as *lists*:

> core_cpus_list:
> 
> human-readable **list** of CPUs within the same core. (deprecated name: “thread_siblings_list”);

Reply:
yes, but again, this is not realistic. On Intel CPUs you can fake whatever topology you'd like and the guest will take it, but on other CPUs (AMD for example) you can't present threads (not 1)
I don't know what will the guest scheduler do with threads > 2. In a real deployment, this is not recommended at all.

Reply:
Keep in mind that exposing threads > 1 will harm the performance of the guest in case these threads (vCPUs) are pinned to pCPUs.